### DISCUSSION TITLE
Problem with CSP Parameter setting from admin console

### DISCUSSION THREAD

**Discussion by nestle2377** [keycloak/keycloak#26793] [Q&A]
Hi to all, today I change the content-security-policy in "security defenses" inside realm setting to add also script-src (like requested from a security audit). 
I added self for script-src.
After restarting keycloak I wasn't able to load the login page of the administration console. there are some inline script I immagine and the browser, with the new directive, doesn't allow me to load the page. 

I've 2 questions:

to roolback and enter the console again was a problem. The only way I find was to update manually the db table whith realm attribute and delete script-src. is there a way to change that parameter without modify the db entry?

which setting you suggest to enter in Content-SecurityPolicy attribute to avoid the login page of the administration console and include also script-src?

thanks!

**Comment by darius-m** [comment]
If the page is broken after changing the CSP, the web interface is likely unusable (at least as long as the browser properly enforces CSP; there may be a way to disable it temporarily, but it may not be a great idea, just in case you forget to enable it again). The CLI [kcadm script](https://www.keycloak.org/docs/latest/server_admin/index.html#using-the-admin-cli) can be used to configure the system from the CLI, but I don't know how it should be used (or if it can be used) for CSP specifically. It's probably worth looking into.

The easiest way to allow the use of inline scripts is to add the `unsafe-inline` keyword. This is, however, overly broad in most cases, and does not really improve security by a lot. As far as I know, the recommended way to handle inline scripts would be to have the backend generate hashes or nonces and send them to the client to mark the page as "safe", but I'm not sure whether Keycloak supports this (I think the scripts would have to be generated dynamically for each request, so nonces are not reused).

You can try reading the [Mozilla docs](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Security-Policy) for more details on the use of each keyword.

**Reply by nestle2377** [reply]
Thanks Darius, I will try with the kcadm script!
